### PR TITLE
eks-prow-build-cluster: Enable EKS Access Entries

### PR DESCRIPTION
**Pull request by xmudrii**
AWS has introduced the [EKS Access Entries](https://docs.aws.amazon.com/eks/latest/userguide/access-entries.html) as a new way to manage access to EKS clusters and as a replacement for the `aws-auth` ConfigMap. Up until now, we have managed the access to the eks-prow-build-cluster via the `aws-auth` ConfigMap.

This PR is the first phase of the migration to the EKS Access Entries for the eks-prow-build-cluster. It includes:

- Upgrading the needed Terraform dependencies/modules to the latest versions
- Minor refactor and cleanup of the Terraform code for the eks-prow-build-cluster to be compatible with the newer module versions
- Switching the authentication mode for the EKS Prow build cluster from `CONFIG_MAP` to `API_AND_CONFIG_MAP`
- Adding the EKS Access Entries that match the `aws-auth` ConfigMap
- Update the IAM permissions to allow managing the EKS Access Entries
- Deploy the eks-pod-identity-agent addon so that Pods can use the EKS Pod Identity API to authenticate with AWS

There'll be at least one more phase after this one that will:

- Remove the `aws-auth` ConfigMap
- Switch the authentication mode from `API_AND_CONFIG_MAP` to `API`

It's also important to note that the newer versions of the EKS Terraform module will stop supporting both `CONFIG_MAP` and `API_AND_CONFIG_MAP` authentication modes, it'll be only possible to use the `API` mode (see https://github.com/terraform-aws-modules/terraform-aws-eks/pull/2858 for more details).

There shouldn't be any significant downtime caused by this migration and the kubeconfig files remain the same, so we can consider that this is a safe migration. There might be a very short downtime while the change is being applied in a way that it's temporary not possible to access the cluster, but that shouldn't last longer than a minute or two.

This change is already applied on the Canary cluster. We'll discuss on Slack plans for applying this to the Production cluster.
/hold

/assign @dims @BenTheElder @upodroid @koksay

### CODE REVIEW

File: infra/aws/terraform/modules/eks-prow-iam/policy_apply.tf [hunk 1]
```diff
@@ -60,6 +60,7 @@ data "aws_iam_policy_document" "eks_apply" {
       "ec2:ModifyInstanceAttribute",
       "ec2:TerminateInstances",
       "ec2:ImportKeyPair",
+      "eks:CreateAccessEntry",
```
Comment (bryantbiggs) [inline infra/aws/terraform/modules/eks-prow-iam/policy_apply.tf:63]: fyi - there is also an `eks:UpdateAccessEntry`

**Review by bryantbiggs — COMMENTED** [review]
looks great! left a few suggestions to consider now or in the future

File: infra/aws/terraform/modules/eks-prow-iam/policy_apply.tf (hunk 1 above)
Comment (xmudrii) [reply infra/aws/terraform/modules/eks-prow-iam/policy_apply.tf:63]: Added, thanks!

File: infra/aws/terraform/prow-build-cluster/data.tf [hunk 2]
```diff
@@ -18,6 +18,10 @@ data "aws_caller_identity" "current" {}
 
 data "aws_availability_zones" "available" {}
 
+data "aws_iam_role" "sso_admins" {
```
[the remark this answers is not in the document]
Comment (xmudrii) [reply infra/aws/terraform/prow-build-cluster/data.tf:21]: I was looking for exactly this, but couldn't find the solution! Thanks a lot, updated!

File: infra/aws/terraform/prow-build-cluster/locals.tf [hunk 3]
```diff
@@ -41,6 +41,54 @@ locals {
 
   azs = slice(data.aws_availability_zones.available.names, 0, 3)
 
+  default_access_entries = {
+    # Admin entries
+    eks-infra-admin = {
```
[the remark this answers is not in the document]
Comment (xmudrii) [reply infra/aws/terraform/prow-build-cluster/locals.tf:46]: To make the migration easier and safer, we'll use our policies, but we'll look into migrating to the managed policies.

**Comment by dims** [comment]
thanks @xmudrii @bryantbiggs 

/approve
/lgtm

**Comment by xmudrii** [comment]
The migration has been completed successfully 🎉 

```
Apply complete! Resources: 14 added, 6 changed, 3 destroyed.
```

@upodroid @ameukam I had to do some quick fixes before the migration because Terraform was complaining, can you please LGTM the PR again?

/hold cancel

**Review by upodroid — APPROVED** [review]

**Pull request merged**